### PR TITLE
use cached PropertyID everywhere + Refactor

### PR DESCRIPTION
**Pull request by float3**

### CODE REVIEW

**Comment by pema99** [comment]
lmk if you plan to revisit this. currently cursed

**Comment by float3** [comment]
I'm gonna run benchmarks this weekend

**Comment by float3** [comment]
on my machine
1000 property sets by name take about 4.5 milliseconds
1000 property sets by id take about 1.7 milliseconds
while 
1000 PropertyToID method calls take about 4.1 milliseconds (these would happen only on Start())

File: AudioLinkUnityProject/Packages/.gitignore [hunk 1]
```diff
@@ -1,2 +1,3 @@
 com.vrchat.*/
 !com.vrchat.core.*/
+/*/
```
Comment (pema99) [inline AudioLinkUnityProject/Packages/.gitignore:3]: wtf are you doing

File: AudioLinkUnityProject/Packages/com.vrchat.core.vpm-resolver/package.json [hunk 2]
```diff
@@ -1,17 +1,17 @@
...
+	"author" : {
+		"name" : "VRChat",
+		"email" : "developer@vrchat.com",
+		"url" : "https://github.com/vrchat/packages"
+	},
+	"url" : "",
+	"dependencies" : {
+		"com.unity.nuget.newtonsoft-json" : "2.0.2"
+	}
+}
```
Comment (pema99) [inline AudioLinkUnityProject/Packages/com.vrchat.core.vpm-resolver/package.json:17]: I mean, this is fine.. but please with the unrelated diffs in your PR's x)

File: Packages/com.llealloo.audiolink.extras/Runtime/AudioLink.Extras.USharp.asset [hunk 3]
```diff
@@ -0,0 +1,16 @@
+%YAML 1.1
```
Comment (pema99) [inline Packages/com.llealloo.audiolink.extras/Runtime/AudioLink.Extras.USharp.asset:1]: what is going on with the diff here - I swear to god I pushed this file to dev already

File: AudioLinkUnityProject/Packages/.gitignore (hunk 1 above)
Comment (float3) [reply AudioLinkUnityProject/Packages/.gitignore:3]: I cherrypicked the resolver update commit onto this branch because I needed/wanted to use newer U# and not have to keep manually swapping to it everytime I switched to this branch

File: Packages/com.llealloo.audiolink.extras/Runtime/Scripts/AudioLinkZone.asset [hunk 4]
```diff
@@ -12,15 +12,16 @@ MonoBehaviour:
   m_Script: {fileID: 11500000, guid: c333ccfdd0cbdbc4ca30cef2dd6e6b9b, type: 3}
```
Comment (pema99) [inline Packages/com.llealloo.audiolink.extras/Runtime/Scripts/AudioLinkZone.asset:12]: Is your diff fucked or is this file also intentional? It's hard to tell

File: Packages/com.llealloo.audiolink/Editor/Scripts/AudioLinkConfiguratorEditor.cs [hunk 5]
```diff
@@ -8,7 +8,7 @@
 #else
   using UnityEditor.Experimental.SceneManagement;
 #endif
-#if UDON
+#if UDONSHARP
```
Comment (pema99) [inline Packages/com.llealloo.audiolink/Editor/Scripts/AudioLinkConfiguratorEditor.cs:11]: I don't remember this define. Can you tell me when it is defined? I only remember UDON, COMPILER_UDONSHARP and UNITY_EDITOR really

File: Packages/com.llealloo.audiolink/Runtime/Scripts/AudioLink.cs [hunk 6]
```diff
@@ -124,10 +126,120 @@ public class AudioLink : MonoBehaviour
...
+        //Raw audio data.
+        internal int _Samples0L;
+        internal int _Samples1L;
+        internal int _Samples2L;
+        internal int _Samples3L;
+        
+        internal int _Samples0R;
+        internal int _Samples1R;
+        internal int _Samples2R;
+        internal int _Samples3R;
```
Comment (pema99) [inline Packages/com.llealloo.audiolink/Runtime/Scripts/AudioLink.cs:184]: any particular reason these are internal?

File: Packages/com.llealloo.audiolink.extras/Runtime/AudioLink.Extras.USharp.asset (hunk 3 above)
Comment (float3) [reply Packages/com.llealloo.audiolink.extras/Runtime/AudioLink.Extras.USharp.asset:1]: Again I cherrypicked the commit because the errors were annoying me everytime I switched to this branch, please check the commit history of this pr

File: Packages/com.llealloo.audiolink/Editor/Scripts/AudioLinkConfiguratorEditor.cs (hunk 5 above)
Comment (float3) [reply Packages/com.llealloo.audiolink/Editor/Scripts/AudioLinkConfiguratorEditor.cs:11]: UDONSHARP is sorta new and it's set by U#, this is probably unnecessary with VPM but it's more correct  than UDON for our usecase because UDON can be true without U# being imported into the project

File: Packages/com.llealloo.audiolink/Runtime/Scripts/AudioReactiveSurface.cs [hunk 7]
```diff
@@ -2,10 +2,13 @@
 
 namespace VRCAudioLink
 {
-    #if UDON
+    #if UDONSHARP
         using UdonSharp;
 
         public class AudioReactiveSurface : UdonSharpBehaviour
+    #else
```
Comment (pema99) [inline Packages/com.llealloo.audiolink/Runtime/Scripts/AudioReactiveSurface.cs:9]: nice!

File: Packages/com.llealloo.audiolink/Runtime/Scripts/AudioLink.cs [hunk 8]
```diff
@@ -313,14 +424,14 @@ private void Update()
             }
 
             // use _AdvancedTimeProps.w for Debugging
-            audioMaterial.SetVector("_AdvancedTimeProps", new Vector4(
+            audioMaterial.SetVector(_AdvancedTimeProps0, new Vector4(
```
[the remark this answers is not in the document]
Comment (float3) [reply Packages/com.llealloo.audiolink/Runtime/Scripts/AudioLink.cs:427]: I already grepped the docs, didn't find any mention

File: Packages/com.llealloo.audiolink/Runtime/Scripts/AudioLinkController.cs [hunk 9]
```diff
@@ -161,14 +161,14 @@ public void UpdateSettings()
...
 
-                audioSpectrumDisplay.SetFloat("_X0", x0Slider.value);
-                audioSpectrumDisplay.SetFloat("_X1", x1Slider.value);
-                audioSpectrumDisplay.SetFloat("_X2", x2Slider.value);
-                audioSpectrumDisplay.SetFloat("_X3", x3Slider.value);
-                audioSpectrumDisplay.SetFloat("_Threshold0", threshold0Slider.value);
-                audioSpectrumDisplay.SetFloat("_Threshold1", threshold1Slider.value);
-                audioSpectrumDisplay.SetFloat("_Threshold2", threshold2Slider.value);
-                audioSpectrumDisplay.SetFloat("_Threshold3", threshold3Slider.value);
+                audioSpectrumDisplay.SetFloat(audioLink._X0, x0Slider.value);
```
[the remark this answers is not in the document]
Comment (float3) [reply Packages/com.llealloo.audiolink/Runtime/Scripts/AudioLinkController.cs:164]: Yes

File: Packages/com.llealloo.audiolink.extras/Runtime/Scripts/AudioLinkZone.asset (hunk 4 above)
Comment (float3) [reply Packages/com.llealloo.audiolink.extras/Runtime/Scripts/AudioLinkZone.asset:12]: another cherrypick

File: AudioLinkUnityProject/Packages/com.vrchat.core.vpm-resolver/package.json (hunk 2 above)
Comment (float3) [reply AudioLinkUnityProject/Packages/com.vrchat.core.vpm-resolver/package.json:17]: sorry, but the manual updating was really annoying me

File: Packages/com.llealloo.audiolink/Runtime/Scripts/AudioLink.cs (hunk 6 above)
Comment (float3) [reply Packages/com.llealloo.audiolink/Runtime/Scripts/AudioLink.cs:184]: they don't all have to be internal only like a handful but I wanted to be consistent and I don't see any downside

File: AudioLinkUnityProject/Packages/.gitignore (hunk 1 above)
Comment (pema99) [reply AudioLinkUnityProject/Packages/.gitignore:3]: what does that have to do with the gitignore? This doesn't seem like something that should be pushed...
btw, you can ignore files locally. might wanna look into that

File: Packages/com.llealloo.audiolink/Editor/Scripts/AudioLinkConfiguratorEditor.cs (hunk 5 above)
Comment (pema99) [reply Packages/com.llealloo.audiolink/Editor/Scripts/AudioLinkConfiguratorEditor.cs:11]: Set during compilation or just set when U# is imported?

File: Packages/com.llealloo.audiolink/Runtime/Scripts/AudioLinkController.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (pema99) [reply Packages/com.llealloo.audiolink/Runtime/Scripts/AudioLinkController.cs:164]: Unresolving so we have somewhere to talk about this: I suppose we should check that this is actually worth it given the overhead (probably?) of getting a field from another udon behavior. I don't know, maybe you could even check by just compiling and looking at the uasm - it might be obvious

File: Packages/com.llealloo.audiolink/Editor/Scripts/AudioLinkConfiguratorEditor.cs (hunk 5 above)
Comment (float3) [reply Packages/com.llealloo.audiolink/Editor/Scripts/AudioLinkConfiguratorEditor.cs:11]: when U# is imported

File: Packages/com.llealloo.audiolink/Runtime/Scripts/AudioLinkController.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (float3) [reply Packages/com.llealloo.audiolink/Runtime/Scripts/AudioLinkController.cs:164]: it's one extern 
`EXTERN,"VRCUdonCommonInterfacesIUdonEventReceiver.__GetProgramVariable__SystemString__SystemObject"`
plus one extern to cast it `EXTERN, "SystemConvert.__ToInt32__SystemObject__SystemInt32"`

[the remark this answers is not in the document]
Comment (float3) [reply Packages/com.llealloo.audiolink/Runtime/Scripts/AudioLinkController.cs:164]: Ok I've decided to have each Behaviour keep track of it's own IDs, unless you have any objections I will go ahead with that

[the remark this answers is not in the document]
Comment (pema99) [reply Packages/com.llealloo.audiolink/Runtime/Scripts/AudioLinkController.cs:164]: just make the fields not internal anymore then

[the remark this answers is not in the document]
Comment (float3) [reply Packages/com.llealloo.audiolink/Runtime/Scripts/AudioLinkController.cs:164]: already done, just haven't pushed yet

File: Packages/com.llealloo.audiolink/Runtime/Scripts/AudioLink.cs [hunk 10]
```diff
@@ -72,16 +74,16 @@ public class AudioLink : MonoBehaviour
...
         #endif
         public string customString1;
-        #if UDON
+        #if UDONSHARP
         [UdonSynced]
         #endif
         public string customString2;
 
-        [Header("Internal (Do not modify)")] public Material audioMaterial;
+        [Header("private (Do not modify)")] public Material audioMaterial;
```
Comment (pema99) [inline Packages/com.llealloo.audiolink/Runtime/Scripts/AudioLink.cs:86]: Bruh moment

File: Packages/com.llealloo.audiolink/Runtime/Scripts/ThemeColorController.cs [hunk 11]
```diff
@@ -11,10 +11,21 @@ namespace VRCAudioLink
...
+            #if UDONSHARP
+            [UdonSynced] 
+            #endif
+            
+            private int themeColorMode;
+
+            #if UDONSHARP
+            [UdonSynced] 
+            #endif
+
```
Comment (pema99) [inline Packages/com.llealloo.audiolink/Runtime/Scripts/ThemeColorController.cs:27]: bad newline

Comment (pema99) [reply Packages/com.llealloo.audiolink/Runtime/Scripts/ThemeColorController.cs:27]: nbd
